modules/siddhi-core/src/main/java/io/siddhi/core/table/CacheTable.java

```diff
@@ -0,0 +1,305 @@
+
+        // initialize cache table
+        MetaStreamEvent cacheTableMetaStreamEvent = new MetaStreamEvent();
+        cacheTableMetaStreamEvent.addInputDefinition(cacheTableDefinition);
+        for (Attribute attribute : cacheTableDefinition.getAttributeList()) {
+            cacheTableMetaStreamEvent.addOutputData(attribute);
+        }
+
+        StreamEventFactory cacheTableStreamEventFactory = new StreamEventFactory(cacheTableMetaStreamEvent);
+        StreamEventCloner cacheTableStreamEventCloner = new StreamEventCloner(cacheTableMetaStreamEvent,
+                cacheTableStreamEventFactory);
+        super.initTable(cacheTableDefinition, cacheTableStreamEventFactory,
+                cacheTableStreamEventCloner, configReader, siddhiAppContext, recordTableHandler);
+    }
+
+    public void addStreamEventUptoMaxSize(StreamEvent streamEvent) {
+        int sizeAfterAdding = this.size();
+        ComplexEventChunk<StreamEvent> addEventsLimitCopy = new ComplexEventChunk<>(true);
+        do {
+            sizeAfterAdding++;
+            addEventsLimitCopy.add((StreamEvent) generateEventWithRequiredFields(streamEvent, siddhiAppContext,
+                    cacheExpiryEnabled));
+            streamEvent = streamEvent.getNext();
+        } while (sizeAfterAdding != maxSize && streamEvent != null);
+        readWriteLock.writeLock().lock();
+        try {
+            this.add(addEventsLimitCopy);
+        } finally {
+            readWriteLock.writeLock().unlock();
+        }
+    }
+
+    public void addAndTrimUptoMaxSize(ComplexEventChunk<StreamEvent> addingEventChunk) {
+        ComplexEventChunk<StreamEvent> addingEventChunkForCache = new ComplexEventChunk<>(true);
+        addingEventChunk.reset();
+        while (addingEventChunk.hasNext()) {
+            StreamEvent event = addingEventChunk.next();
+            addingEventChunkForCache.add((StreamEvent) generateEventWithRequiredFields(event, siddhiAppContext,
+                    cacheExpiryEnabled));
+        }
+        readWriteLock.writeLock().lock();
+        try {
+            super.add(addingEventChunkForCache);
+            if (this.size() > maxSize) {
+                this.deleteEntriesUsingCachePolicy(this.size() - maxSize);
+            }
+        } finally {
+            readWriteLock.writeLock().unlock();
+        }
+    }
+
+    public void updateOrAddAndTrimUptoMaxSize(ComplexEventChunk<StateEvent> updateOrAddingEventChunk,
+                                              CompiledCondition compiledCondition,
+                                              CompiledUpdateSet compiledUpdateSet,
+                                              AddingStreamEventExtractor addingStreamEventExtractor, int maxTableSize) {
+        ComplexEventChunk<StateEvent> updateOrAddingEventChunkForCache = new ComplexEventChunk<>(true);
+        updateOrAddingEventChunk.reset();
+        while (updateOrAddingEventChunk.hasNext()) {
```

Comment:
For all ComplexEventChunk cases events should be first removed from the iterated chunk or cloned, before they are added to the new chunk

Reply:
generateEventWithRequiredFields() function creates a new complexEvent based on the ComplexEvent from the iterated chunk. i.e cloned with some modification. So iterated chunk and new chunk are not using the same events